### DISCUSSION TITLE
Is there a way to define a custom "encoder" when model_dump-ing to dict?

### DISCUSSION THREAD

**Discussion by dmarkowvw** [pydantic/pydantic#9275] [Question]
Given this model
```python
class MyModel(BaseModel):
    my_field: AwareDatetime

m = MyModel(my_field=datetime.now(UTC))
```
when I run
```pyhton
m.model_dump()
```
instead of it returning a dict containing a datetime object for `my_field` I'd like to have a way to cast that field to string such that the return will be the following dict

```python
{"my_field" : "2024-01-01T22:21:22.234344+00:00"}
```

Similar to defining a custom encoder in the model for model_dump_json() but for dumping a Python dict.

Background: Using DynamoDB with FastAPI and Pydantic one has to think up nasty workarounds to be able to put_item using the boto3 resource when the Python dict contains a datetime or float object.

**Comment by giancarloromeo** [comment]
You might try the following (in _Pydantic v2_):

```python
import datetime as dt

from pydantic import AwareDatetime, BaseModel


class MyModel(BaseModel):
    my_field: AwareDatetime

m = MyModel(my_field=dt.datetime.now(dt.timezone.utc))
m.model_dump(mode="json")
```

```
{'my_field': '2024-09-23T11:39:31.738414Z'}
```

See: https://docs.pydantic.dev/2.9/api/base_model/#pydantic.BaseModel.model_dump

**Comment by CharlesPerrotMinot** [comment]
So, there used to be (with pydantic v1), but the short answer is no, there's no "easy" way to do this with pydantic v2. We're using a wrapping `field_serializer`, `@field_serializer("*", mode="wrap")`, for those types on our `Model` class that we use as the parent class for all our models.


```
@field_serializer("*", mode="wrap")
def serialize_special_types(  # type: ignore[no-untyped-def] # noqa: ANN201
    self,
    value: Any,
    nxt: SerializerFunctionWrapHandler,
    _info: SerializationInfo,
):
    if isinstance(value, datetime):
        return isoformat(value)
    if isinstance(value, Decimal):
        return str(value)
    return nxt(value)
```
